csrc/preseg_passes/mark_aliases_prepare.cpp

```diff
@@ -49,62 +51,82 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
 
-  for (TensorView* aliased_out : aliased_outs) {
-    // Rarely, if `aliased_out` is already defined by `segment_set`, don't
-    // create another `segment_set`.
-    if (LoadStoreOp* def =
-            dynamic_cast<LoadStoreOp*>(aliased_out->definition())) {
-      if (def != nullptr && def->opType() == LoadStoreOpType::SegmenterSet) {
+  // Mark all expressions that are (transitively) used by non-alias outputs.
+  auto used_by_non_aliases =
+      [](const std::vector<Expr*>& exprs) -> std::unordered_set<Expr*> {
+    return {exprs.begin(), exprs.end()};
+  }(StmtSort::getExprsTo(non_alias_outs));
+
+  for (TensorView* aliased_io : aliased_ios) {
+    // Divide the users of aliased_io into two groups according to
+    // `used_by_non_aliases`.
+    std::vector<Expr*> users_used_by_non_aliases;
+    std::vector<Expr*> users_used_only_by_aliases;
+    for (Expr* e : aliased_io->uses()) {
+      if (used_by_non_aliases.count(e)) {
+        users_used_by_non_aliases.push_back(e);
+      } else {
+        users_used_only_by_aliases.push_back(e);
+      }
+    }
+
+    // If all users are used by non-alias outputs, do nothing. Adding
+    // segment_set around unlikely creates more no-op regions.
+    if (users_used_only_by_aliases.empty()) {
+      continue;
+    }
+
+    // If all users are used by aliases, put a `segment_set` before it.
+    if (users_used_by_non_aliases.empty()) {
+      if (aliased_io->isFusionInput()) {
+        // A `segment_set` before a fusion input is useless.
         continue;
       }
+      // Rarely, if `aliased_io` is already defined by `segment_set`, don't
+      // create another `segment_set`.
+      if (LoadStoreOp* def =
+              dynamic_cast<LoadStoreOp*>(aliased_io->definition())) {
+        if (def->opType() == LoadStoreOpType::SegmenterSet) {
+          continue;
+        }
+      }
+      aliased_io->cacheBefore(LoadStoreOpType::SegmenterSet);
+      continue;
     }
 
-    // This is suboptimal in many uncommon cases. My head hurts when thinking
-    // about them, so go simple for now :)
-    //
-    // Legend:
-    //   M* = a meta op defining a **fusion output**
-    //   N/M = a non-meta op defining a **fusion output**
-    //
-    // Case 1:
-    //
-    //   N/M -> N/M
-    //      |
-    //      -->  M
-    //
-    // We should put a `segment_set` on the **edge** from N/M to M, so the two
-    // `N/M`s go to the same kernel.
-    //
-    // Case 2:
-    //
-    //   N/M -> M1 -> M2
-    //           |
-    //           --> N/M
-    //
-    // We should change it to
-    //
-    //   N/M -> M1 -> M2
-    //      |
-    //      --> M1' (non-output copy of M1) -> N/M
-    //
-    // and then put a `segment_set` on N/M->M1.
-    aliased_out->cacheBefore(LoadStoreOpType::SegmenterSet);
+    // Some users of `aliased_io` are used by non-aliases, and some are used
+    // only by aliases. We put a `segment_set` after `aliased_io` and redirect
+    // the latter group to use the `segment_set`.
```

Comment:
qq: if aliased_io is a fusion input, shouldn't we skip this as well?

Reply:
I believe this is still needed for fusion inputs and AliasTest.SegmentMetaOps is an example. There, a fusion input is used by a permute on one branch and a series of compute op on the other branch. The code here adds a segment_set between the fusion input and the permute, separating the permute and the compute.

Reply:
I see... The segmenter set is necessary for the segmenter logic.